BootloaderCorePkg/Library/MpInitLib/Ia32/MpInitLib.c

```diff
@@ -74,16 +75,18 @@ CpuInit (
 
   if (PcdGet8 (PcdSmmRebaseMode) == SMM_REBASE_ENABLE_ON_S3_RESUME_ONLY) {
     if (GetBootMode() == BOOT_ON_S3_RESUME) {
-      SmmBaseInfo = (SMMBASE_INFO *)PcdGet32 (PcdSmramTsegBase);
-      if (SmmBaseInfo->Signature == PLD_TO_LDR_SMM_SIG) {
-        for (CpuIdx = 0; CpuIdx < SmmBaseInfo->CpuEntry; CpuIdx++) {
+      S3SaveReg = (S3_SAVE_REG *) PcdGet32 (PcdSmramTsegBase);
+      SmmBaseInfo = (SMMBASE_INFO *) ((UINT8 *)S3SaveReg + sizeof(S3_SAVE_REG) + sizeof(REG_INFO) * S3SaveReg->RegCnt);
+      SmmBaseInfo = (SMMBASE_INFO *) ALIGN_UP ((UINT32)SmmBaseInfo, sizeof(UINT32));
+      if (SmmBaseInfo->Signature == SMM_BASE_INFO_SIG) {
+        for (CpuIdx = 0; CpuIdx < SmmBaseInfo->CpuCnt; CpuIdx++) {
           if (ApicId == SmmBaseInfo->SmmBase[CpuIdx].ApicId) {
             SmmRebase (Index, ApicId, SmmBaseInfo->SmmBase[CpuIdx].SmmBase);
             break;
           }
         }
-        if (CpuIdx == SmmBaseInfo->CpuEntry) {
-          ASSERT (CpuIdx < SmmBaseInfo->CpuEntry);
+        if (CpuIdx == SmmBaseInfo->CpuCnt) {
+          ASSERT (CpuIdx < SmmBaseInfo->CpuCnt);
         }
```

Comment:
Looks it will never ASSERT since it runs to it only when CpuIdx == SmmBaseInfo->CpuCnt.
do you mean ASSERT (FALSE)?

Reply:
It can ASSERT when the ApicId for a proc is missing in the ApicIds reported via TSEG region by the payload.

Reply:
Looks there is duplication there are two conditions.
remove If condition or remove condition inside ASSERT would be more clear.